File: concatenation-of-array/Solution1929.java
// https://leetcode.com/problems/concatenation-of-array/

import java.util.Arrays;

class Solution1929 {

    // T: O(n), S: O(n)
    public int[] getConcatenation(int[] nums) {
        int n = nums.length;
        var ans = new int[n * 2];
        for (int i = 0; i < n; i++) {
            ans[i] = nums[i];
            ans[i + n] = nums[i];
        }
        return ans;
    }

    // T: O(n), S: O(n)
    public int[] getConcatenationBuiltin(int[] nums) {
        int n = nums.length;
        var ans = new int[n * 2];
        System.arraycopy(nums, 0, ans, 0, n);
        System.arraycopy(nums, 0, ans, n, n);
        return ans;
    }

    public static void main(String[] args) {
        var sol = new Solution1929();
        int[][] testCases = {
                { 1, 2, 1 },
                { 1, 3, 2, 1 },
                { 1, 2, 3, 4 }
        };
        for (int[] nums : testCases) {
            var ans = sol.getConcatenation(nums);
            System.out.println(Arrays.toString(ans));
        }
        System.out.println(Arrays.toString(sol.getConcatenationBuiltin(new int[] { 1, 2, 3 })));
    }

}

File: concatenation-of-array/solution1929.go
// https://leetcode.com/problems/concatenation-of-array/

package main

import "fmt"

// T: O(n), S: O(n)
func getConcatenation(nums []int) []int {
	n := len(nums)
	ans := make([]int, 2*n)

	for i := 0; i < n; i++ {
		ans[i] = nums[i]
		ans[i+n] = nums[i]
	}

	return ans
}

// T: O(n), S: O(n)
func getConcatenationBuiltin(nums []int) []int {
	return append(nums, nums...)
}

func main() {
	testCases := [][]int{
		{1, 2, 1},
		{1, 3, 2, 1},
		{1, 2, 3, 4},
	}
	for _, testCase := range testCases {
		fmt.Println(getConcatenation(testCase))
	}
	fmt.Println(getConcatenationBuiltin([]int{1, 2, 1}))
}
